docs/DesignDocs/full_edl_serialization.md

```diff
@@ -0,0 +1,151 @@
+Enforcing full serialization in EDL
+===================================
+
+Motivation
+----------
+
+Currently EDL only provides partial support for parameter serialization. For
+example, the following EDL defines ECALLs whose parameters can only be
+partially serialized.
+
+```c++
+enclave
+{
+    /* Include struct iovec (foreign structure) */
+    #include <sys/uio.h>
+
+    trusted
+    {
+        public int send(
+            [in, count=iovcnt] const struct iovec* iov,
+            size_t iovcnt);
+
+        public int recv(
+            [out, count=iovcnt] struct iovec* iov,
+            size_t iovcnt);
+    };
+};
+```
+
+The generator produces code that serializes the array of **iovec**
+structures, but it cannot serialize the buffer referenced by the structure
+itself. Recall the definition of the **iovec** structure.
+
+```c++
+struct iovec
+{
+    void* iov_base;
+    size_t iov_len;
+};
+```
+
+Consequently, these ECALLs pass pointers to untrusted memory (**iov_base**) to
+the enclave developer's **send** and **recv** implementations.
+
+The exposure of untrusted pointers to the enclave implementation poses two main
+problems.
+
+- Implementations are vulnerable to coding errors that may introduce security
+  vulnerabilities.
+- The enclave application is not portable to trusted hardware environments
+  that do not support or have not enabled shared memory (e.g., TrustZone).
+
+Some potential security vulnerabilities include:
+
+- Time-of-check, Time-of-use errors (TOC-TOU).
+- Unwittingly writing secrets onto host memory.
+- Overwrite attacks where **iov_base** points to enclave memory.
+
+EDL provides a *deep-copy* feature for fully serializing parameters as shown
+below.
+
+```c++
+enclave
+{
+    struct local_iovec
+    {
+        [size=iov_len]
+        void* iov_base;
+        size_t iov_len;
+    };
+
+    trusted
+    {
+        public int send(
+            [in, count=iovcnt] const struct local_iovec* iov,
+            size_t iovcnt);
+
+        public int recv(
+            [out, count=iovcnt] struct local_iovec* iov,
+            size_t iovcnt);
+    };
+};
+```
+
+This example introduces a local structure definition (**local_iovec**) which
+provides an annotation allowing **iov_base** to be serialized.
+
+The following EDL considers a second scenario that implicitly passes untrusted
+memory into the trusted implementation.
+
+```c++
+enclave
+{
+    untrusted
+    {
+        struct widget* get_widget();
```

Comment:
this might be a bad example, since you could potentially implement this by assuming [out] for all return pointers.  A stronger example might be a variable sized result like an array or buffer, e.g.

struct widget* get_widgets([out] int* widget_count);